external/llvm-project/mlir/lib/Conversion/AsyncToGPU/AsyncToGPU.cpp

```diff
@@ -0,0 +1,354 @@
+//===- AsyncToGPU.cpp - Convert Async to GPU dialect ----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/AsyncToGPU/AsyncToGPU.h"
+
+#include "../PassDetail.h"
+#include "mlir/Dialect/Async/IR/Async.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Interfaces/CallInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#define DEBUG_TYPE "convert-async-to-gpu"
+
+using namespace mlir;
+using namespace mlir::async;
+
+//===----------------------------------------------------------------------===//
+// Convert Async dialect types to GPU types.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// AsyncGPUTypeConverter only converts types from the Async dialect to
+/// the corresponding GPU type and does not convert any other types.
+class AsyncGPUTypeConverter : public TypeConverter {
+public:
+  AsyncGPUTypeConverter() {
+    addConversion([](Type type) { return type; });
+    addConversion(convertAsyncTypes);
```

Comment:
That is, why isn't this `addConversion([](TokenType type) { return gpu::AsyncTokenType::get(type.getContext()); });`